### PR TITLE
Demo aggregate core apis.

### LINKED ISSUE #2848: [kubeapps-apis] Demo core.v1alpha1.GetPackageRepositories
### Description:

The current demo functionality, such as `GetPackageRepositories` and `GetAvailablePackages`, are exposed per plugin, for example:

```
$ curl -s http://localhost:8080/plugins/helm_fluxv2/packages/v1alpha1/packagerepositories | jq .
{
  "repositories": [
    {
      "name": "bitnami",
      "namespace": "flux-system",
      "url": "https://charts.bitnami.com/bitnami"
    }
  ]
}

$ curl -s http://localhost:8080/plugins/kapp_controller/packages/v1alpha1/packagerepositories | jq .
{
  "repositories": [
    {
      "name": "repo-name.example.com",
      "namespace": "",
      "url": "foo.registry.example.com/repo-name/main@sha256:cecd9b51b1f29a773a5228fe04faec121c9fbd2969de55b0c3804269a1d57aa5"
    }
  ]
}
```

This allows clients to interact with specific plugins when needed, as we cannot assume that this will not be necessary in some circumstances, such as a plugin-specific UX using extra features of that format.

That said, we also want to ensure it's possible to provide a generic implementation of the packages API implemented in our core, that returns results from all plugins, such as:

```
$ curl -s http://localhost:8080/core/packages/v1alpha1/packagerepositories | jq .
{
  "repositories": [
    {
      "name": "bitnami",
      "namespace": "flux-system",
      "url": "https://charts.bitnami.com/bitnami",
      "plugin": "fluxv2"
    },
    {
      "name": "repo-name.example.com",
      "namespace": "",
      "url": "foo.registry.example.com/repo-name/main@sha256:cecd9b51b1f29a773a5228fe04faec121c9fbd2969de55b0c3804269a1d57aa5",
      "plugin": "kapp-controller"
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by absoludity**
### Description of the change

This was an investigative PR that adds the ability to recognise when plugins satisfy the interface for core services (well, currently only core.packaging.v1alpha1) and register them for aggregation in the core api, so that you can:

```
curl -s http://localhost:8080/core/packages/v1alpha1/packagerepositories | jq .
{
  "repositories": [
    {
      "name": "bitnami",
      "namespace": "flux-system",
      "url": "https://charts.bitnami.com/bitnami",
      "plugin": {
        "name": "fluxv2.packages",
        "version": "v1alpha1"
      }
    },
    {
      "name": "demo-package-repository",
      "url": "k8slt/corp-com-pkg-repo:1.0.0",
      "plugin": {
        "name": "kapp_controller.packages",
        "version": "v1alpha1"
      }
    }
  ]
}
```
and similar for available packages.

It demonstrates only those two calls at the moment, but the idea is that a client could then refer to such an available package and hit the core.packages api to install it, and the core api can call the relevant plugin implementation etc., enabling a client to potentially treat all plugins in a homogeneous way, where possible, while still allowing each plugin to have custom extra functionality for clients to use as needed.

### Benefits

The nice thing is that we can do this via go's interfaces in the same way that grpc does so.

### Applicable issues

  - fixes #2848

### Additional information

I have an idea how I can test the core API functions, but will do that in a follow-up PR if that's OK.

I'll add some notes in context inline.

### CODE REVIEW

File: cmd/kubeapps-apis/buf.gen.yaml
Comment (absoludity) [inline cmd/kubeapps-apis/buf.gen.yaml:25]: I'll add an upstream issue and link here on Monday.

File: cmd/kubeapps-apis/server/packages.go
Comment (absoludity) [inline cmd/kubeapps-apis/server/packages.go:23]: This is the specially generated struct which we embed as always, we're just not forced to do so with the `require_unimplemented_servers=false`, and since the required interface is then generated *without* the private method, plugin servers are able to satisfy the interface as long as they implement the actual functions (eg. `GetPackageRepositories`, `GetAvailablePackages`, ... whatever else we add).

File: cmd/kubeapps-apis/server/packages.go
Comment (absoludity) [inline cmd/kubeapps-apis/server/packages.go:40]: So there were a number of different ways to call the separate plugin versions. The simplest would have been to use the client and actually do the grpc call to localhost, but that then requires serializing/deserializing for every request. Instead, because we recorded at startup which plugins satisfy the core packaging server interface and store a reference to the implementation, we can just use the implementations directly here without going over the wire at all, so no serialization/deserialization required.

File: cmd/kubeapps-apis/server/plugins.go
Comment (absoludity) [inline cmd/kubeapps-apis/server/plugins.go:74]: This file is refactored a little so that the plugin detail can be passed down the chain of calls and included together with the implementation server to use in the core server.

File: cmd/kubeapps-apis/server/plugins.go
Comment (absoludity) [inline cmd/kubeapps-apis/server/plugins.go:143]: This is the new piece that is called when registering a grpc function ^^^ , defined below.

File: cmd/kubeapps-apis/server/plugins.go
Comment (absoludity) [inline cmd/kubeapps-apis/server/plugins.go:159]: This error should not be possible, since we've already checked that the pluginSrv does indeed implement the core packages interface, but I can't ignore it :)

File: cmd/kubeapps-apis/buf.gen.yaml
Comment (antgamdia) [reply cmd/kubeapps-apis/buf.gen.yaml:25]: So, just for clarification: there is nothing wrong in continue using `require_unimplemented_servers=true` if the `mustEmbedUnimplemented**Server` was exposed when generating the code.
So, the upstream issue is for them to expose it, am I getting it right?

**Review by antgamdia — APPROVED** [review]
Thank you for this interesting approach for registering the plugins satisfying the interface!

File: cmd/kubeapps-apis/buf.gen.yaml
Comment (absoludity) [reply cmd/kubeapps-apis/buf.gen.yaml:25]: Woops, still need to create the upstream issue. Yes, that's what I'll be suggesting upstream, though there may be a reason I'm not aware of as to why they want it private.

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/main.go
Comment (absoludity) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/main.go:29]: Indeed, thanks.

File: cmd/kubeapps-apis/server/packages.go
Comment (absoludity) [reply cmd/kubeapps-apis/server/packages.go:41]: Yeah - especially since it's stored as (a pointer to) the interface, makes it trivial here.

File: cmd/kubeapps-apis/server/plugins.go
Comment (absoludity) [reply cmd/kubeapps-apis/server/plugins.go:120]: `registerGRPC` was only updated to a method of the `pluginServer` because I needed to store state while registering each plugin. Yep, I could update `registerHTTP` to be a method as well for consistency, but my preference is for stateless functions where possible (easier to reason about and obvious that there's no state of pluginServer being used).

File: cmd/kubeapps-apis/server/server.go
Comment (absoludity) [reply cmd/kubeapps-apis/server/server.go:70]: Yes, as in, the kubeapps-apis service manually registers both the `core.plugins` and `core.packages` packages, the only two core packages currently. Yes, the `core.plugins` package also loads and registers all the loaded plugins.

I actually want to do another PR to move the `core.plugins` and `core.packages` into their own (versioned) packages too, but I'll wait until your auth PR has landed before doing that refactor.

**Pull request merged**
